bench/nbody.cc

```diff
@@ -0,0 +1,266 @@
+{
+    Timer<time_unit> timer;
+    unsigned long long t[kNruns];
+    for (size_t n = 0; n < kNruns; n++) {
+        timer.Start();
+
+        for (std::size_t i = 0; i < kN; i++) {
+            for (std::size_t j = 0; j < kN; j += VectorSize<Float_v>()) {
+                __m256 p1velx = _mm256_load_ps(&velx[j]);
+                __m256 p1vely = _mm256_load_ps(&vely[j]);
+                __m256 p1velz = _mm256_load_ps(&velz[j]);
+                pPInteractionAVX2(
+                    _mm256_load_ps(&posx[j]),
+                    _mm256_load_ps(&posy[j]),
+                    _mm256_load_ps(&posz[j]),
+                    p1velx,
+                    p1vely,
+                    p1velz,
+                    _mm256_broadcast_ss(&posx[i]),
+                    _mm256_broadcast_ss(&posy[i]),
+                    _mm256_broadcast_ss(&posz[i]),
+                    _mm256_broadcast_ss(&mass[i])
+                );
+                _mm256_store_ps(&velx[j], p1velx);
+                _mm256_store_ps(&vely[j], p1vely);
+                _mm256_store_ps(&velz[j], p1velz);
+            }
+        }
+        for (std::size_t i = 0; i < kN; i += VectorSize<Float_v>()) {
+            _mm256_store_ps(&posx[i], _mm256_fmadd_ps(_mm256_load_ps(&velx[i]), vTimestep, _mm256_load_ps(&posx[i])));
+            _mm256_store_ps(&posy[i], _mm256_fmadd_ps(_mm256_load_ps(&vely[i]), vTimestep, _mm256_load_ps(&posy[i])));
+            _mm256_store_ps(&posz[i], _mm256_fmadd_ps(_mm256_load_ps(&velz[i]), vTimestep, _mm256_load_ps(&posz[i])));
+        }
+
+        t[n] = timer.Elapsed();
+    }
+
+    const unsigned long long mean = std::accumulate(std::begin(t), std::end(t), 0ull) / (kN * kNruns);
+    printf("%20s %6llu\n", "AVX2 Intrinsics", mean);
+}
+#endif
+
+template <class Backend>
+inline void pPInteractionSIMD(
+    typename Backend::Float_v p1posx,
+    typename Backend::Float_v p1posy,
+    typename Backend::Float_v p1posz,
+    typename Backend::Float_v& p1velx,
+    typename Backend::Float_v& p1vely,
+    typename Backend::Float_v& p1velz,
+    typename Backend::Float_v p2posx,
+    typename Backend::Float_v p2posy,
+    typename Backend::Float_v p2posz,
+    typename Backend::Float_v p2mass)
+{
+    const typename Backend::Float_v xdistance = p1posx - p2posx;
+    const typename Backend::Float_v ydistance = p1posy - p2posy;
+    const typename Backend::Float_v zdistance = p1posz - p2posz;
+    const typename Backend::Float_v xdistanceSqr = xdistance * xdistance;
+    const typename Backend::Float_v ydistanceSqr = ydistance * ydistance;
+    const typename Backend::Float_v zdistanceSqr = zdistance * zdistance;
+    const typename Backend::Float_v distSqr = eps2 + xdistanceSqr + ydistanceSqr + zdistanceSqr;
+    const typename Backend::Float_v distSixth = distSqr * distSqr * distSqr;
+    const typename Backend::Float_v invDistCube = 1.0f / math::Sqrt(distSixth);
+    const typename Backend::Float_v sts = p2mass * invDistCube * timeStep;
```

Comment:
Rather than template on the full backend, you can template only on the actual types you need, i.e. on the float type only. That will make syntax a bit more palatable. You can see the mandelbrot benchmark for an example of what I'm suggesting. The quadratic benchmark is older, that's why it uses this style.

Reply:
Good point!